<h4 style="text-align: center;"> Curso: Programación para el análisis de datos</h4><br>
    <h4 style="text-align: center;">Tarea: Unidad 1 - ejercicio 2</h4><br>
<h4 style="text-align: center;">Presentado por: Oscar Andrés Russi Porras</h4><br>
<h4 style="text-align: center;">Grupo: 10</h4><br>
<h4 style="text-align: center;">Código: 1018436442</h4><br>
<h4 style="text-align: center;">Presentado a: Felipe Alexander Pipicano Guzmán</h4><br>
<h4 style="text-align: center;">Universidad Nacional Abierta y a Distancia – UNAD</h4><br>
<h4 style="text-align: center;">Fecha: 15/09/2023</h4><br>

<h3>Solución de los ejercicios</h3><br><br>
<lu><li>construya la siguiente matriz como un arreglo de numpy:
    $$
A =
\begin{bmatrix}
13 & 1 \\
14 & 1 \\
15 & 1 \\
16 & 1 \\
17 & 1 \\
18 & 1 \\
\end{bmatrix}
$$
    </li></lu>

In [20]:
import numpy as np
A = np.array([[13, 1], [14, 1], [15, 1], [16, 1], [17, 1], [18, 1]])
print(A)

[[13  1]
 [14  1]
 [15  1]
 [16  1]
 [17  1]
 [18  1]]


<ul><li>Construya el siguiente vector como arreglo de numpy:</li></ul>
$$
b =
\begin{bmatrix}
16 \\
24 \\
45 \\
57 \\
75 \\
102 \\
\end{bmatrix}
$$

In [21]:
b = np.array([16,24,45,57,75,102])
print(b)

[ 16  24  45  57  75 102]


<ul><li>Calcule el siguiente vector x, usando operaciones de matrices (dot, transpose, inv) de numpy.linalg:
   $$
x=(A^{T}A)^{-1}A^{T}b
$$ 
    </li></ul>

In [22]:
transpose_A= A.T # transpose of A
TA_dot_A = np.dot(transpose_A, A) # multiply tranpose of A and A
inv_TA_dot_A =  np.linalg.inv(TA_dot_A) # inverse previous result
TA_dot_b = np.dot(transpose_A, b) # multiply tranpose of A and b
x_matrix  = np.dot(inv_TA_dot_A, TA_dot_b)
print(x_matrix)


[  17.         -210.33333333]


<ul><li>y calcule ‖𝐴𝑥−𝑏‖ utilizando la función numpy.linalg.norm.</li></ul>

In [23]:
A_x_dot = np.dot(A, x_matrix) # multiply A and x matrix
# print(A_x_dot)
diff_A_x_dot_And_b = A_x_dot - b # get matrix difference
norm_final_matrix = np.linalg.norm(diff_A_x_dot_And_b) # get norm of the previous matrix
print(norm_final_matrix)

10.83205120618128


<ul><li>Solucione el siguiente sistema de ecuaciones, mediante la función numpy.linalg.solve (de ser necesario, investigue la función en la documentación de numpy):
   $$
(A^{T}A)x=A^{T}b
$$ 
    </li></ul>

In [24]:
x_solution = np.linalg.solve(TA_dot_A, TA_dot_b)
print(x_solution)

[  17.         -210.33333333]


<ul><li>y calcule ‖𝐴𝑥−𝑏‖ utilizando la función numpy.linalg.norm.</li></ul>

In [25]:
norm_final2 = np.linalg.norm(np.dot(A,x_solution)-b) # calculate again norm using the matrix calculated in the previous cell
print(norm_final2)

10.832051206181271


<ul><li>Solucione el siguiente sistema de ecuaciones, mediante la función numpy.linalg.lstsq (de ser necesario, investigue la función en la documentación de numpy):
   $$
Ax=b
$$ 
    </li></ul>

In [26]:
solved_system = np.linalg.lstsq(A, b, rcond=None) # solve system Ax = b
print(solved_system)

(array([  17.        , -210.33333333]), array([117.33333333]), 2, array([38.27438209,  0.26772348]))


<ul><li>y calcule ‖𝐴𝑥−𝑏‖ utilizando la función numpy.linalg.norm.</li></ul>

In [27]:
norm_final3 = np.linalg.norm(np.dot(A,solved_system[0])-b) # calculate again norm using the matrix calculated in the previous cell
print(norm_final3)

10.832051206181276


# Conclusiones

<ul>
    <li>
    La siguiente es una propiedad de la multiplicación de matrices $$ A X A^{-1} = I $$ 
donde $$ I $$ es la matriz identidad. Y además $$ A X I = A $$ 

por lo anterior al tener la siguiente operación: $$ x=(A^{T}A)^{-1}A^{T}b $$
si multiplicamos a ambos lados por el valor: $$ (A^{T}A) $$
obtenemos lo siguiente  $$ (A^{T}A)x=A^{T}b $$
De forma que en ambas ecuaciones, el vector o matriz x tienen el mismo valor.
    </li><br>
    <li>
        según el principio de mínimos cuadrados, si la ecuación de matrices $$ Ax=b $$ no tiene solución se puede obtener una solución aproximada con la siguiente fórmula:
       $$
        x \approx (A^{T}A)^{-1}A^{T}b
       $$
        Pero en caso de que la ecuación tenga solución entonces el método de mínimos cuadrados permitirá llegar a la solución exacta. De forma que si Ax=b tiene solución entonces:
        $$
        A^{-1}b = (A^{T}A)^{-1}A^{T}b
        $$
    </li><br>
    <li>
        en el desarrollo del ejercicio, el método lstsq de numpy retornó la siquiente tupla:
    (array([  17.        , -210.33333333]), array([117.33333333]), 2, array([38.27438209,  0.26772348]))
        1) La solución, esto es el vector que se obtiene mediante el método de mínimos cuadrados. En este caso la solución es el array  [17, -210.33333333] que es el mismo valor obtenido con el método solve.
        2) Residuos, esto es la suma de las distancias entre los valores esperados y los valores observados. En este caso el valor es 117.33333333. Residuos es un array porque lstsq permite operar sobre múltiples vectores.
        3) Esto es el rango de la matriz, esto es el número de filas linealmente independientes en la matriz, en este caso el valor es 2.
        4) 
    </li><br>
    <li>
    La transpuesta de una matriz(expresada usualmente como A^t) donde las columnas son equivalentes a las filas de 
la matriz original.<br>
En consecuencia cada valor de la columna original cuyas coordenadas estaban en las coordenadas ij pasa a estar en las 
coordenadas ji.<br>
Ejemplo, si tenemos la matriz:
$$
  \begin{bmatrix}
1 & 2 \\
3 & 4 \\
\end{bmatrix}
$$
Su transpuesta es: 
$$
  \begin{bmatrix}
1 & 3 \\
2 & 4 \\
\end{bmatrix}
$$
    </li><br>
    <li>
    
la operación punto entre dos matrices (expresada como A*B) es un tipo de multiplicación de matrices donde el resultado final
es una nueva matriz con i filas (número de filas de la segunda matriz) y j columnas (número de columnas de la primera matriz)
y donde el valor de cada valor en coordenadas ij es la suma del producto de los valores de la fila i en A y de la columna 
j en B. Para poder hacer la operación punto la primera matriz debe tener un número de filas equivalente a las columnas de 
la segunda matriz .
ejemplo dee multiplicación de matrices:
$$
\begin{bmatrix}
1 & 2 \\
3 & 4 \\
\end{bmatrix}
\cdot
\begin{bmatrix}
5 \\
6 \\
\end{bmatrix}
$$

Basado en la explicación anteriormente dada el resultado de esta multiplicación sería:
$$
\begin{bmatrix}
1 \times 5 + 2 \times 6 \\
3 \times 5 + 4 \times 6 \\
\end{bmatrix}
= 
\begin{bmatrix}
17 \\
39 \\
\end{bmatrix}
$$
    </li><br>
    <li>
    La matriz invertida se construye de la siguiente forma, teniendo una matriz de n x n, realizar operaciones sobre esta hasta obtener la matriz identidad, realizar las mismas operaciones sobre una matriz identidad y el resultado final de esta segunda matriz es la invertida. Ejemplo:
$$
\begin{bmatrix}
1 & 2 & 1 \\
2 & 3 & 1 \\
1 & 1 & 2 \\
\end{bmatrix}
\begin{bmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 0 & 1 \\
\end{bmatrix}
$$
empezamos a construir la matriz triangular superior en la primera columna:
$$
\begin{bmatrix}
1 & 2 & 1 \\
0 & -1 & -1 \\
0 & -1 & 1 \\
\end{bmatrix}
\begin{bmatrix}
1 & 0 & 0 \\
-2 & 1 & 0 \\
-1 & 0 & 1 \\
\end{bmatrix}
$$
seguimos construyendo la matriz triangular con la segunda columna:
$$
\begin{bmatrix}
1 & 2 & 1 \\
0 & -1 & -1 \\
0 & 0 & 2 \\
\end{bmatrix}
\begin{bmatrix}
1 & 0 & 0 \\
-2 & 1 & 0 \\
1 & -1 & 1 \\
\end{bmatrix}
$$
empezamos a construir la matriz identidad en la última file:
$$
\begin{bmatrix}
1 & 2 & 0 \\
0 & -1 & 0 \\
0 & 0 & 1 \\
\end{bmatrix}
\begin{bmatrix}
\frac{1}{2} & \frac{1}{2} & \frac{-1}{2} \\
\frac{-3}{2} & \frac{1}{2} & \frac{1}{2} \\
\frac{1}{2} & \frac{-1}{2} & \frac{1}{2} \\
\end{bmatrix}
$$
continuamos con la primera fila:
$$
\begin{bmatrix}
1 & 0 & 0 \\
0 & -1 & 0 \\
0 & 0 & 1 \\
\end{bmatrix}
\begin{bmatrix}
\frac{-5}{2} & \frac{3}{2} & \frac{1}{2} \\
\frac{-3}{2} & \frac{1}{2} & \frac{1}{2} \\
\frac{1}{2} & \frac{-1}{2} & \frac{1}{2} \\
\end{bmatrix}
$$
terminamos de hacer la matriz identidad en la segunda fila:
$$
\begin{bmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 0 & 1 \\
\end{bmatrix}
\begin{bmatrix}
\frac{-5}{2} & \frac{3}{2} & \frac{1}{2} \\
\frac{3}{2} & \frac{-1}{2} & \frac{-1}{2} \\
\frac{1}{2} & \frac{-1}{2} & \frac{1}{2} \\
\end{bmatrix}
$$
    </li><br>
    <li>
    Teniendo dos matrices A y B donde ambas tienen las mismas dimensiones ij, la suma o diferencia es el resultado de sumar (o restar) cada par de valores de A y B que comparten las mismas coordenadas. Ejemplo:

$$
\begin{bmatrix}
1 & 2 \\
3 & 4 \\
\end{bmatrix}
+
\begin{bmatrix}
5 & 6 \\
7 & 8 \\
\end{bmatrix}
=
\begin{bmatrix}
5+1 & 6+2 \\
7+3 & 8+4 \\
\end{bmatrix}
=
\begin{bmatrix}
6 & 8 \\
10 & 12 \\
\end{bmatrix}
$$

</li><br>
    <li>   
La norma de un vector o una matriz es la raiz cuadrada de los cuadrados de los componentes de la matriz o vector. Ejemplo, teniendo la siguiente matriz:
$$
\begin{bmatrix}
5 & 1 \\
7 & 5 \\
\end{bmatrix}
$$
la norma de esta matriz es:
$$
\sqrt{5^2 + 1^2 + 5^2 + 7^2} = \sqrt{25 + 1 + 25 + 49} = \sqrt{100} = 10
$$
    </li><br>
    <li>
    En términos algebráicos se puede expresar que resolver un sistema lineal de ecuaciones e encontrar x de la función:

$$
Ax=b
$$

Donde A es una matriz, x es la columna de las incógnitas y b es la columna vector dee términos independientes. Existen muchos métodos para encontrar la solución, una forma es invirtiendo la matriz, mediante sustitución,y también mediante el método crammer. En este último método vamos sustituyendo una de las columnas de A por la del vector b, encontramos la determinante de esta nueva matriz y la dividimos por la determinante de la matriz A, y repetimos para cada columna y así econtramos todas las incógnitas. Ejemplo, teniendo el siguiente sistema de ecuaciones:

$$
2x + y = 7
\\
x + 3y = 6
$$
para x tenemos que:
$$
\det 
\begin{bmatrix}
7 & 1 \\
6 & 3 \\
\end{bmatrix}
\div
\det
\begin{bmatrix}
2 & 1 \\
1 & 3 \\
\end{bmatrix}
= 
\frac{7 \times 3 - 1 \times 6}{2 \times 3 - 1 \times 1} 
= 
\frac{15}{5}
=
3
$$
Y para y tenemos:
$$
\det 
\begin{bmatrix}
2 & 7 \\
1 & 6 \\
\end{bmatrix}
\div
\det
\begin{bmatrix}
2 & 1 \\
1 & 3 \\
\end{bmatrix}
= 
\frac{2 \times 6 - 7 \times 1}{2 \times 3 - 1 \times 1} 
= 
\frac{5}{5}
=
1
$$
por lo tanto:
$$
x=3, y=1
$$
    </li><br>
    <li>numpy.array es un método que permite almacenar arreglos multidimensionales, entendiendo por arreglos una estructura de datos que puede ser usada para representar vectores, matrices y tensores. Ejemplo: "numpy.Array([[1,2,][3,4]])".
    </li><br>
    <li>numpy.dot es un método que permite multiplicar numpy arrays (cuando la operación es posible según las reglas de la multiplicación de matrices). Ejemplo: "numpy.dot(numpy.array([1,2]),numpy.array([3,4]))".
    </li><br>
    <li>El método T de numpy.array permite obtener la transpuesta de una matriz.Ejemplo de uso: "numpy.Array([[1,2,][3,4]]).T".
    </li><br>
    <li>
        El método numpy.linalg.inv permite obtener la inversa de una matriz cuadrada (que son el único tipo de matrices invertibles). Un ejemplo de su uso sería: "numpy.linalg.inv(numpy.array([[1,2],[3,4]]))".
    </li><br>
    <li>
        El método numpy.linalg.norm permite calcular la norma de una matriz. Ejemplo: "numpy.linalg.norm(numpy.array([[1,2],[3,4]]))".
    </li><br>
    <li>
        Con el método numpy.linalg.solve permite resolver ecuaciones de algebra lineal de la forma Ax=b. Ejemplo, teniendo la siguient ecuación:
        $$
        \begin{bmatrix}
1 & 2 \\
3 & 4 \\
\end{bmatrix}x=
\begin{bmatrix}
2 & 4 \\
1 & 1 \\
\end{bmatrix}
        $$
        podemos encontrar x con
        "numpy.linalg.solve(numpy.array([[1,2],[3,4]]),numpy.array([[2,4],[1,1]]))".
    </li><br>
    <li>
            Con el método numpy.linalg.lstsq podemos encontrar la aproximación de los mínimos cuadrados de la ecuación de algebra lineal en la forma Ax-b. Ejemplo, teniendo la siguiente ecuación:
        $$
        \begin{bmatrix}
1 & 2 \\
3 & 4 \\
\end{bmatrix}x-
\begin{bmatrix}
2 \\
4 \\
\end{bmatrix}
        $$
        podemos encontrar x con "numpy.linalg.lstsq(numpy.array([[1,2],[3,4]]),numpy.array([2,4]))"
        
    </li><br>
</ul>

<h3>Bibliografía</h3>
<ul>
<li>Alex Galea. (2018). Applied Data Science with Python and Jupyter: Use Powerful Industry-standard Tools to Unlock New, Actionable Insights From Your Data: Vol. 1st edition.</li>
<li>Boschetti, A., & Massaron, L. (2016). Python Data Science Essentials - Second Edition (Vol. 0002).</li>
<li>Samir Madhavan. (2015). Mastering Python for Data Science: Explore the World of Data Science Through Python and Learn How to Make Sense of Data.</li>
<li>Thakur, A. (2016). Python: Real-World Data Science. Packt Publishing. (pp. 491 - 508).</li>
<li>McKinney, W. (2015). Pandas, python data analysis library. URL http://pandas. pydata. org, 3-15.</li>
</ul>

In [30]:
import numpy as np
np.linalg.lstsq(np.array([[1,1],[1,1],[2,-1]]),np.array([2,1,1]))

C:\Users\Usuario\AppData\Local\Temp\ipykernel_18676\1084922482.py:2: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  np.linalg.lstsq(np.array([[1,1],[1,1],[2,-1]]),np.array([2,1,1]))


(array([0.83333333, 0.66666667]),
 array([0.5]),
 2,
 array([2.44948974, 1.73205081]))